### PR TITLE
feat: Implement deep input validation

### PR DESCRIPTION
**Pull request by phako**

### CODE REVIEW

**Comment by phako** [comment]
@mhier I have changed the API - it is backward compatible now

**Comment by mhier** [comment]
I think we should call automatically `enableDeepValidation()` when `validate(TransferElementID())` is called (that will validate the initial value). We should also force people to actually do that call by throwing an exception if the first call to `validate()` has a valid (i.e. not default-constructed) `TransferElementID`. The pointer to the owning module can be extracted e.g. from the first accessor passed to the `add()` call (via the slightly ugly `dynamic_cast<Module*>(node.getOwningModule())->findApplicationModule()` call).

File: Modules/src/UserInputValidator.cc [hunk 1]
```diff
@@ -44,6 +61,86 @@ namespace ChimeraTK {
     return rejected;
   }
 
+  void UserInputValidator::enableDeepValidation() {
```
Comment (mhier) [inline Modules/src/UserInputValidator.cc:64]: Please place chrome bar /*****/ between function implementation

File: Modules/include/UserInputValidator.h [hunk 2]
```diff
@@ -137,16 +141,23 @@ namespace ChimeraTK {
     bool validateAll();
 
    protected:
+    static constexpr std::string_view tagValidatedVariable{"__UserInputValidator"};
+
+    // Helper function to set up queue lengths of valid values
+    void enableDeepValidation();
```
Comment (mhier) [inline Modules/include/UserInputValidator.h:147]: Maybe mention in the comment that this function is called in the first call to validate() automatically.

Also maybe rename it back into finalise()

File: Modules/src/UserInputValidator.cc [hunk 3]
```diff
@@ -16,16 +18,31 @@ namespace ChimeraTK {
 
   bool UserInputValidator::validate(const ChimeraTK::TransferElementID& change) {
     if(!change.isValid()) {
+      enableDeepValidation();
```
Comment (mhier) [inline Modules/src/UserInputValidator.cc:21]: I think it is/ was so far ok to call `validateAll()` directly instead of `validate({})`, so maybe better move the call to `enableDeepValidation()` into `validateAll()`.

File: Modules/src/UserInputValidator.cc [hunk 4]
```diff
@@ -44,6 +61,86 @@ namespace ChimeraTK {
...
+        _downstreamInvalidatingReturnChannels.emplace(accessor.getAppAccessorNoType().getId());
+      }
+    }
+
+    auto keepPvAccesWithReturnChannel = Model::EdgeFilter([](const Model::EdgeProperties& edge) -> bool {
+      return edge.type == Model::EdgeProperties::Type::pvAccess && edge.pvAccessWithReturnChannel;
+    });
+
+    // Filter that keeps our tagged variable nodes and application modules
+    auto validatedVariablesAndApplicationModulesFilter =
```
Comment (mhier) [inline Modules/src/UserInputValidator.cc:78]: Why don't you use `keepTag(UserInputValidator::tagValidatedVariable) || keepApplicationModules()` instead?

File: Modules/src/UserInputValidator.cc [hunk 5]
```diff
@@ -44,6 +61,86 @@ namespace ChimeraTK {
...
+  void UserInputValidator::enableDeepValidation() {
+    // Find out accessors with return
+    for(auto& accessor : _module->getAccessorListRecursive()) {
+      if(accessor.getDirection() == VariableDirection{VariableDirection::feeding, true} &&
+          accessor.getModel().getTags().count(std::string(UserInputValidator::tagValidatedVariable)) > 0) {
+        _downstreamInvalidatingReturnChannels.emplace(accessor.getAppAccessorNoType().getId());
+      }
+    }
+
+    auto keepPvAccesWithReturnChannel = Model::EdgeFilter([](const Model::EdgeProperties& edge) -> bool {
```
Comment (mhier) [inline Modules/src/UserInputValidator.cc:73]: This one is a candidate for an official filter, I guess :-)

File: Modules/src/UserInputValidator.cc [hunk 6]
```diff
@@ -44,6 +61,86 @@ namespace ChimeraTK {
...
+
+    auto orderVisitor = [&](auto proxy) {
+      if constexpr(Model::isApplicationModule(proxy)) {
+        stack.push_front(proxy);
+        distances.try_emplace(&proxy.getApplicationModule(), std::numeric_limits<int>::min());
+      }
+    };
+
+    _module->getModel().visit(orderVisitor, Model::visitOrderPost, Model::depthFirstSearch,
+        keepPvAccesWithReturnChannel, validatedVariablesAndApplicationModulesFilter);
```
Comment (mhier) [inline Modules/src/UserInputValidator.cc:106]: I fear this does not do what you intend. `visit()` will always iterate all vertices in the model that are reachable with the given edge filter, independent of the vertex filter. The vertex filter will only control for which found vertices the visitor functor is called. The vertex filter does not prevent the DFS to continue a certain path. Only the edge filter does that.

Since you have `if constexpr(Model::isApplicationModule(proxy))` in your visitor functor, the vertex filter is completely redundant. You can leave out and you will get the same result here.

As a consequence, this DFS will not stop in the hierarchy at the lowest level of validators. As long as there are return channels, it will continue its search.

Maybe this is actually ok, since we only want to estimate the maximum possible validation depth here. Return channels are not so common, and it might actually be that some other validation mechanism is behind the return channel. This code can be quite simplified then though, and we should mention in a comment that this is an approximation.

File: Modules/src/UserInputValidator.cc [hunk 7]
```diff
@@ -44,6 +61,86 @@ namespace ChimeraTK {
...
+    auto downstreamModuleCollector = [&](auto proxy) {
+      if constexpr(Model::isApplicationModule(proxy)) {
+        downstreamModulesWithFeedback.insert(&proxy.getApplicationModule());
+      }
+    };
+
+    auto connectingVariableVisitor = [&](auto proxy) {
+      if constexpr(Model::isVariable(proxy)) {
+        proxy.visit(downstreamModuleCollector, Model::adjacentOutSearch, validatedVariablesAndApplicationModulesFilter,
+            Model::keepApplicationModules);
```
Comment (mhier) [inline Modules/src/UserInputValidator.cc:119]: here you are passing two vertex filters to the same `visit()` call, which is not allowed. The `Model::keepApplicationModules` is redundant, as it is already part of the `validatedVariablesAndApplicationModulesFilter`. I am not 100% sure what this code tries to do, so it's hard to judge whether it is correct... Is it intentional that there is no `keepPvAccessWithReturnChannel`?

File: include/Model.h [hunk 8]
```diff
@@ -697,6 +697,14 @@ namespace ChimeraTK::Model {
 
   struct ReturnFirstHit : SearchOption {};
 
+  struct VisitOrder : SearchOption {
```
Comment (mhier) [inline include/Model.h:700]: It would be good to have a dedicated test for this new feature. Please write a ticket that we need to write one if not yet done...

File: tests/executables_src/testUserInputValidator.cc [hunk 9]
```diff
@@ -620,19 +615,13 @@ BOOST_AUTO_TEST_CASE(testDeepBackwardsPropagation) {
...
-  BOOST_TEST(midstreamIn == 13); // first value is coming from upstream
-  BOOST_TEST(downstrIn.readNonBlocking());
-  BOOST_TEST(downstrIn == 14); // first value is coming from upstream/midstream
 
+  test.stepApplication();
   BOOST_TEST(downstrIn.readLatest()); // just observe final state, because intermediate states might be subject
   BOOST_TEST(downstrIn == 5);         // to race conditions
   BOOST_TEST(midstreamIn.readLatest());
   BOOST_TEST(midstreamIn == 4);
-  BOOST_TEST(upstrIn.readLatest());
```
Comment (mhier) [inline tests/executables_src/testUserInputValidator.cc:633]: Why is this line removed?

File: Modules/src/UserInputValidator.cc (hunk 1 above)
Comment (phako) [reply Modules/src/UserInputValidator.cc]: Fixed

File: Modules/include/UserInputValidator.h (hunk 2 above)
Comment (phako) [reply Modules/include/UserInputValidator.h]: Also fixed

File: Modules/src/UserInputValidator.cc (hunk 5 above)
Comment (phako) [reply Modules/src/UserInputValidator.cc]: Moved to Model.h and added to the list of filters

File: Modules/src/UserInputValidator.cc (hunk 4 above)
Comment (phako) [reply Modules/src/UserInputValidator.cc]: Can be removed completely

File: Modules/src/UserInputValidator.cc (hunk 7 above)
Comment (phako) [reply Modules/src/UserInputValidator.cc]: That should have been keepPvAccessWIthReturnChannel, not validatedVariables... Fixed.

File: tests/executables_src/testUserInputValidator.cc (hunk 9 above)
Comment (phako) [reply tests/executables_src/testUserInputValidator.cc:633]: It failed because there is no update. But I should have replaced it with !readLatest

Comment (phako) [reply tests/executables_src/testUserInputValidator.cc:633]: I ahve added the test with !readLatest

File: Modules/src/UserInputValidator.cc (hunk 3 above)
Comment (phako) [reply Modules/src/UserInputValidator.cc]: Changed

**Review by mhier — APPROVED** [review]

**Comment by mhier** [comment]
Sorry, I screwed up the code formatting correct yesterday. Should I fix this and then merge? Or do you still have open issues to change?

File: include/Model.h (hunk 8 above)
Comment (phako) [reply include/Model.h:701]: Will write a ticket for this

**Comment by phako** [comment]
Need to write a ticket for the new test, but should be done code-wise

File: Modules/src/UserInputValidator.cc [hunk 10]
```diff
@@ -18,14 +20,29 @@ namespace ChimeraTK {
...
       return validateAll();
     }
+
+    if(!_finalised) {
+      throw ChimeraTK::logic_error("Initial values were not validated");
+    }
+
+    // We have downstream channels that signalized a change - invalidate all of our
+    if(_downstreamInvalidatingReturnChannels.count(change) > 0) {
+      _module->setCurrentVersionNumber({});
```
Comment (mhier) [inline Modules/src/UserInputValidator.cc:30]: (Commenting on a long-merged PR)

Do you remember why (or even whether) this line was necessary? It seems to cause a race condition which is even visible in one of the tests...
